### PR TITLE
Rebuild layer if a change in ARG is detected

### PR DESCRIPTION
**Pull request by umohnani8**
Signed-off-by: Urvashi Mohnani <umohnani@redhat.com>

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Check whether the ARG in the containerfile is changed by
either the --build-arg flag or local environment and use
the cached layer or rebuild the layer accordingly.
Add tests for this use case as well.

#### Which issue(s) this PR fixes:

Fixes https://github.com/containers/buildah/issues/2848
Fixes https://github.com/containers/buildah/issues/2837

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Comment by umohnani8** [comment]
@nalind @rhatdan PTAL

**Comment by TomSweeneyRedHat** [comment]
All kinds of test unhappiness @umohnani8

**Comment by rhatdan** [comment]
LGTM

**Comment by rhatdan** [comment]
@containers/podman-maintainers PTAL

File: tests/bud.bats [hunk 1]
```diff
@@ -1716,6 +1717,30 @@ _EOF
   run_buildah bud --signature-policy ${TESTSDIR}/policy.json --layers -t ${target} -f Dockerfile3 --build-arg=PGDATA=/pgdata --build-arg=UID=17122 --build-arg=CODE=/copr/coprs_frontend --build-arg=USERNAME=praiskup ${TESTSDIR}/bud/build-arg
   run_buildah inspect -f '{{.FromImageID}}' ${target}
   expect_output "$argsid"
+
+  # If build-arg is specifified via the command line and is different form the previous cached build, it should not use the cached layers.
```
Comment (TomSweeneyRedHat) [inline tests/bud.bats:1721]: nit
```suggestion
  # If build-arg is specifified via the command line and is different from the previous cached build, it should not use the cached layers.
```

File: tests/bud.bats [hunk 2]
```diff
@@ -1716,6 +1717,30 @@ _EOF
...
+  run_buildah inspect -f '{{.FromImageID}}' test-img-2
+  argsid="$output"
+  [[ "$argsid" != "$initialid" ]]
+
+  # Set the build-arg via an ENV in the local environment and verify that the cached layers are not used
+  export TEST=bar
+  run_buildah bud --signature-policy ${TESTSDIR}/policy.json --layers -t test-img-3 --build-arg TEST -f Dockerfile4 ${TESTSDIR}/bud/build-arg
+  run_buildah inspect -f '{{.FromImageID}}' test-img-3
+  argsid="$output"
+  [[ "$argsid" != "$initialid" ]]
```
Comment (TomSweeneyRedHat) [inline tests/bud.bats:1743]: do we have similar tests without the `--layers` option?

Comment (TomSweeneyRedHat) [reply tests/bud.bats:1743]: not sure that we need them, WDYT @umohnani8 ?

**Comment by TomSweeneyRedHat** [comment]
LGTM
One very small question and one nit of a nit in a code comment.
Once I hear back from @umohnani8 I'll merge or wait for the quick repush before merging.

File: tests/bud.bats (hunk 1 above)
Comment (umohnani8) [reply tests/bud.bats]: Fixed

File: tests/bud.bats (hunk 2 above)
Comment (umohnani8) [reply tests/bud.bats:1743]: I don't think we need them. When building without the `--layers` flag we always go through and build every step so should not be an issue in that case.

**Comment by TomSweeneyRedHat** [comment]
LGTM

**Comment by umohnani8** [comment]
We should backport to 1.19 right?

**Pull request merged**